File: vending.c
/* vending.h */
typedef struct Vending_s * Vending_p;

Vending_p
VendingCreate(void);

void
VendingForget(Vending_p self);

int
VendingAmount(Vending_p self);

int
VendingSalesAmount(Vending_p self);

int
VendingDropIn(Vending_p self, 
	      int money);

int
VendingCancel(Vending_p self);

const char *
VendingItemName(Vending_p self);

int
VendingItemPrice(Vending_p self);

int
VendingItemStock(Vending_p self);

int
VendingCanBuy(Vending_p self);

int
VendingBuy(Vending_p self);

/* test.c */
#include <stdio.h>
#include <string.h>
#include <assert.h>

static int
testBuy(void)
{
  Vending_p v = VendingCreate();

  assert(v);
  assert(VendingAmount(v) == 0);
  assert(VendingSalesAmount(v) == 0);
  assert(! strcmp(VendingItemName(v), "Cola"));
  assert(VendingItemPrice(v) == 120);
  assert(VendingItemStock(v) == 5);
  assert(! VendingCanBuy(v));
  assert(! VendingBuy(v));

  VendingDropIn(v, 100);
  assert(VendingAmount(v) == 100);
  assert(VendingSalesAmount(v) == 0);

  VendingDropIn(v, 100);
  assert(VendingAmount(v) == 200);
  assert(VendingCanBuy(v));

  assert(VendingBuy(v));
  assert(VendingSalesAmount(v) == 120);

  assert(VendingCancel(v) == 80);
  assert(VendingAmount(v) == 0);

  VendingDropIn(v, 1000);
  assert(VendingBuy(v));
  assert(VendingBuy(v));
  assert(VendingBuy(v));
  assert(VendingBuy(v));
  assert(! VendingBuy(v));
  assert(VendingAmount(v) == 520);
  assert(VendingSalesAmount(v) == 600);

  VendingForget(v);
  return 1;
}

static int
testDropIn(void)
{
  Vending_p v = VendingCreate();

  assert(! VendingDropIn(v, 1));
  assert(! VendingDropIn(v, 5));
  assert(VendingDropIn(v, 10));
  assert(VendingDropIn(v, 50));
  assert(VendingDropIn(v, 100));
  assert(VendingDropIn(v, 500));
  assert(VendingDropIn(v, 1000));
  assert(VendingAmount(v) == 1660);

  return 1;
}

int
main(int argc,
     char **argv)
{
  assert(testBuy());
  assert(testDropIn());
  puts("done");
}

/* vending.c */
#include <stdlib.h>

typedef struct {
  const char * name;
  int price;
  int stock;
} Item_t;

typedef struct Vending_s {
  int amount;
  int sales_amount;
  Item_t item;
} Vending_t;

Vending_p
VendingCreate(void)
{
  Vending_p self = calloc(1, sizeof(Vending_t));
  self->item.name = "Cola";
  self->item.price = 120;
  self->item.stock = 5;
  
  return self;
}

void
VendingForget(Vending_p self)
{
  free(self);
}

int
VendingAmount(Vending_p self)
{
  return self->amount;
}

int
VendingSalesAmount(Vending_p self)
{
  return self->sales_amount;
}

int
VendingDropIn(Vending_p self,
	      int money)
{
  static int s_accept[] = {10, 50, 100, 500, 1000, 0};
  int i;
  
  for (i = 0; s_accept[i]; i++) {
    if (s_accept[i] == money) {
      self->amount += money;
      return 1;
    }
  }
  
  return 0;
}

int
VendingCancel(Vending_p self)
{
  int amount = self->amount;
  self->amount = 0;
  return amount;
}

const char *
VendingItemName(Vending_p self)
{
  return self->item.name;
}

int
VendingItemPrice(Vending_p self)
{
  return self->item.price;
}

int
VendingItemStock(Vending_p self)
{
  return self->item.stock;
}

int
VendingCanBuy(Vending_p self)
{
  return self->item.stock >= 1 && self->item.price <= self->amount;
}

int
VendingBuy(Vending_p self)
{
  if (! VendingCanBuy(self)) return 0;

  self->item.stock -= 1;
  self->amount -= self->item.price;
  self->sales_amount += self->item.price;

  return 1;
}

File: vending.rb
require 'test/unit'

class Vending
  Item = Struct.new(:price, :stock, :name)

  def initialize
    @amount = 0
    @sales_amount = 0
    @item = Item.new
    @item.price = 120
    @item.stock = 5
    @item.name = 'Cola'
  end
  attr_reader :amount, :sales_amount, :item

  def drop_in(money)
    return false unless [10, 50, 100, 500, 1000].include?(money)
    @amount += money
    true
  end

  def can_buy?
    @amount >= @item.price && @item.stock > 0
  end

  def buy
    return false unless can_buy?
    @amount -= @item.price
    @sales_amount += @item.price
    @item.stock -= 1
    true
  end

  def cancel
    @amount
  ensure
    @amount = 0
  end
end

class TestVending < Test::Unit::TestCase
  def setup
    @v = Vending.new
  end

  def test_drop_in
    assert(! @v.drop_in(1))
    assert(! @v.drop_in(5))
    assert(@v.drop_in(10))
    assert(@v.drop_in(50))
    assert(@v.drop_in(100))
    assert(@v.drop_in(500))
    assert(@v.drop_in(1000))
    assert_equal(1660, @v.amount)
  end

  def test_buy
    assert_equal(0, @v.amount)
    assert_equal(0, @v.sales_amount)
    assert_equal(120, @v.item.price)
    assert_equal(5, @v.item.stock)
    assert_equal('Cola', @v.item.name)

    assert(! @v.buy)
    assert_equal(120, @v.item.price)
    assert_equal(5, @v.item.stock)
    assert_equal('Cola', @v.item.name)

    @v.drop_in(10)
    @v.drop_in(10)
    assert(! @v.can_buy?)

    @v.drop_in(100)
    assert_equal(120, @v.amount)
    assert(@v.can_buy?)
    assert_equal(0, @v.sales_amount)

    assert(@v.buy)
    assert_equal(0, @v.amount)
    assert_equal(120, @v.sales_amount)
    
    @v.drop_in(1000)
    assert(@v.can_buy?)
    assert(@v.buy)
    assert(@v.buy)
    assert(@v.buy)
    assert(@v.buy)
    assert(! @v.buy)
    assert_equal(600, @v.sales_amount)

    assert_equal(520, @v.amount)
    assert_equal(520, @v.cancel)
    assert_equal(0, @v.amount)
    assert_equal(0, @v.cancel)
  end
end
